pkg/controllers/scheduling/scheduling_controller.go

```diff
@@ -0,0 +1,410 @@
+// 1) The feasible clusters must be from clusterset bound to the placement namespace;
+// 2) If spec.clusterSets is specified, the feasible clusters must belong to any of them;
+// 3. If spec.predicates is specified,the feasible clusters must match any of them;
+func getFeasibleClusters(
+	placement *clusterapiv1alpha1.Placement,
+	clusterLister clusterlisterv1.ManagedClusterLister,
+	clusterSetLister clusterlisterv1alpha1.ManagedClusterSetLister,
+	clusterSetBindingLister clusterlisterv1alpha1.ManagedClusterSetBindingLister) ([]*clusterapiv1.ManagedCluster, error) {
+	// get all clusterset bindings under the placement namespace
+	bindings, err := clusterSetBindingLister.ManagedClusterSetBindings(placement.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	if len(bindings) == 0 {
+		return nil, nil
+	}
+
+	clusterSetNames := sets.NewString()
+	for _, binding := range bindings {
+		// ignore clusterset does not exist
+		_, err := clusterSetLister.Get(binding.Name)
+		if errors.IsNotFound(err) {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		clusterSetNames.Insert(binding.Name)
+	}
+	if len(placement.Spec.ClusterSets) != 0 {
+		clusterSetNames = clusterSetNames.Intersection(sets.NewString(placement.Spec.ClusterSets...))
+	}
+
+	if len(clusterSetNames) == 0 {
+		return nil, nil
+	}
+
+	// list clusters from particular clustersets
+	requirement, err := labels.NewRequirement(clusterSetLabel, selection.In, clusterSetNames.List())
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := labels.NewSelector().Add(*requirement)
+	clusters, err := clusterLister.List(labelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// match clusters with cluster predicates
+	return matchWithClusterPredicates(placement.Spec.Predicates, clusters)
+}
+
+// matchWithClusterPredicates returns a slice of clusters. Each of them must match at least one of the predicates.
+func matchWithClusterPredicates(predicates []clusterapiv1alpha1.ClusterPredicate, clusters []*clusterapiv1.ManagedCluster) ([]*clusterapiv1.ManagedCluster, error) {
+	if len(predicates) == 0 {
+		return clusters, nil
+	}
+	if len(clusters) == 0 {
+		return clusters, nil
+	}
+
+	// prebuild label/claim selectors for each predicate
+	predicateSelectors := [][]labels.Selector{}
+	for _, predicate := range predicates {
+		// build label selector
+		labelSelector, err := convertLabelSelector(predicate.RequiredClusterSelector.LabelSelector)
+		if err != nil {
+			return nil, err
+		}
+		// build claim selector
+		claimSelector, err := convertClaimSelector(predicate.RequiredClusterSelector.ClaimSelector)
+		if err != nil {
+			return nil, err
+		}
+		predicateSelectors = append(predicateSelectors, []labels.Selector{labelSelector, claimSelector})
+	}
+
+	// match cluster with selectors one by one
+	matched := []*clusterapiv1.ManagedCluster{}
+	for _, cluster := range clusters {
```

Comment:
This looks pretty expensive.  It's ok to start, but if this gets expensive, you could try to precompute some indexes.